### PR TITLE
Feature/multiple hierarchies

### PR DESCRIPTION
**Pull request by havelaer**
Issue: #2451 

## What I did

Extended the hierarchy functionality to enable multiple hierarchies next to the main (nameless hierarchy).
I added an extra rootSeparator (not sure about the name though):
```
setOptions({
  // ... 
  hierarchySeparator: /\./,
  rootSeparator: /\|/, // default is false
});
```

**UPDATED**: `rootSeparator` is renamed to `hierarchyRootSeparator` with default value `null`

Extending the API as such:
```
storiesOf('Addons|Addon Info.React Docgen', module)
  .add(          ^
  ...
```

Works backwards compatible with or without setting the hierarchySeparator.

Documentation is updated in addons/options/README.md.

Allowing multiple hierarchies, see issue #2451 for screenshot.

## How to test

I added the rootSeparator to the kitchen sink example and added the "|" separator to the addon stories. 

See cra-kitchen-sink for an example.

The tests are fixed and extended. Though 1 test is failing, but that it is failing on master as well:
```
FAIL  examples/cra-kitchen-sink/src/storyshots.test.js
  ● Test suite failed to run

    TypeError: (0 , _react3.storiesOf)(...).add(...).add(...).add(...).addWithInfo is not a function
```

### CODE REVIEW

File: examples/cra-kitchen-sink/src/stories/addon-a11y.stories.js
Comment (Hypnosphi) [inline examples/cra-kitchen-sink/src/stories/addon-a11y.stories.js]: Looks like now we can drop the "Addon" word everywhere:
`'Addons|A11y'`

**Comment by Hypnosphi** [comment]
> 1 test is failing, but that it is failing on master as well

That's weird. Did you run `yarn bootstrap --core`?

**Comment by Hypnosphi** [comment]
Can you please resolve the conflicts? They are mostly caused by renaming "Left panel" to "Stories panel"

**Comment by Havelaer** [comment]
@Hypnosphi Thanks for reviewing! 
The branch was based on master. I've rebased the branch on latest release/3.3 and fixed the conflicts. 
yarn bootstrap reset everything and all tests are passing!
I've also omitted 'Addon' in the addon story names.

File: lib/ui/src/modules/ui/containers/stories_panel.js
Comment (igor-dv) [inline lib/ui/src/modules/ui/containers/stories_panel.js]: I don't understand why exporting the `resolveStoryRoots` ?  You can hide it under the `resolveStoryHierarchy` method.

File: lib/ui/src/modules/ui/libs/hierarchy.js
Comment (igor-dv) [inline lib/ui/src/modules/ui/libs/hierarchy.js]: Looks like you export this function only for tests, IMO it's not a good thing.

File: lib/ui/src/modules/ui/components/stories_panel/sub_header.js
Comment (Hypnosphi) [inline lib/ui/src/modules/ui/components/stories_panel/sub_header.js]: Looks like it's worth uncommenting. Uppercase text is more readable when it has some letter-spacing

**Review by Hypnosphi — CHANGES_REQUESTED** [review]

**Comment by Hypnosphi** [comment]
Please run `yarn test --core --update` to update test snapshots

File: lib/ui/src/modules/ui/libs/hierarchy.js
Comment (igor-dv) [inline lib/ui/src/modules/ui/libs/hierarchy.js]: I think in this method better return an object, instead of relying on indices in the array.

**Comment by Hypnosphi** [comment]
[Live example](https://deploy-preview-2452--storybooks-cra.netlify.com/?selectedKind=Welcome&selectedStory=to%20Storybook&full=0&addons=1&stories=1&panelRight=1&addonPanel=storybook%2Factions%2Factions-panel) for reference

Looks like some of the stories still have wrong kinds:
image

**Comment by Hypnosphi** [comment]
Root title is hidden when using search. Is it on purpose?
image

File: lib/ui/src/modules/ui/libs/hierarchy.js
Comment (igor-dv) [inline lib/ui/src/modules/ui/libs/hierarchy.js]: `Object.values(rootMap)` ?

File: lib/ui/src/modules/ui/components/stories_panel/sub_header.js
Comment (igor-dv) [inline lib/ui/src/modules/ui/components/stories_panel/sub_header.js]: what do you think about changing `name` to `children` ? I think it would be nicer for use. And I think it should be `required`.

File: lib/ui/src/modules/ui/components/stories_panel/index.js
Comment (igor-dv) [inline lib/ui/src/modules/ui/components/stories_panel/index.js]: worth extracting it to function already.

**Review by igor-dv — CHANGES_REQUESTED** [review]
1. Very like this ability. Would like to use for separating core component in my project.
2. I didn't check the tests yet 
3. I think overloading the storyKind to have more special syntax is a bit wrong direction, but currently, we have nothing to do with it. (maybe create some kind of URI syntax for story) @ndelangen , you have plans for refactoring, WDYT?

File: lib/ui/src/modules/ui/libs/hierarchy.js
Comment (Hypnosphi) [reply lib/ui/src/modules/ui/libs/hierarchy.js]: It's not the main js file, so I think it's Ok to have test-only exports for more granular unit tests

File: lib/ui/src/modules/ui/components/stories_panel/sub_header.js
Comment (Havelaer) [reply lib/ui/src/modules/ui/components/stories_panel/sub_header.js]: Changed it to letterSpacing: '1.2px'

**Comment by Havelaer** [comment]
> Looks like some of the stories still have wrong kinds:

Fixed!

> Root title is hidden when using search. Is it on purpose?

Nice catch! Fixed!

File: addons/options/README.md
Comment (Havelaer) [reply addons/options/README.md]: done!

File: addons/options/src/preview/index.js
Comment (Havelaer) [reply addons/options/src/preview/index.js]: Done! Called it withRegexProp()

File: lib/ui/src/modules/ui/components/stories_panel/index.js
Comment (Havelaer) [reply lib/ui/src/modules/ui/components/stories_panel/index.js]: extracted the map it to separate render method.
extracted the sub header to stories_tree/ as `tree_header`

File: lib/ui/src/modules/ui/components/stories_panel/sub_header.js
Comment (Havelaer) [reply lib/ui/src/modules/ui/components/stories_panel/sub_header.js]: Agreed! Initially copied stories_panel/header.js props interface. But I prefer children prop in such situations. So changed it.

File: lib/ui/src/modules/ui/containers/stories_panel.js
Comment (Havelaer) [reply lib/ui/src/modules/ui/containers/stories_panel.js]: `resolveStoryRoots` is exported for 2 reasons:
1) It's also needed in the stories_panel container.
2) It's has separate unit tests.

Btw I renamed `resolveStoryRoots` to `resolveStoryHierarchyRoots`.

File: lib/ui/src/modules/ui/libs/hierarchy.js
Comment (Havelaer) [reply lib/ui/src/modules/ui/libs/hierarchy.js]: Agreed. Changed it, so it returns an object.

Btw I renamed resolveStoryRoots to resolveStoryHierarchyRoots.

File: lib/ui/src/modules/ui/libs/hierarchy.js
Comment (Havelaer) [reply lib/ui/src/modules/ui/libs/hierarchy.js]: Polyfills take care of `Object.values`?

Code is changed a bit at those lines though: I had to make sure that the main hierarchy is always the first in the array, so I didn't have use for `Object.values` anymore.

Which triggers the next question:
What could we do with the sorting of the other hierarchies in the panel? Now it respects the order in which the stories are loaded.

**Review by Hypnosphi — APPROVED** [review]
LGTM. Let's merge it right after the upcoming `3.3.0` release

File: lib/ui/src/modules/ui/libs/hierarchy.js
Comment (igor-dv) [inline lib/ui/src/modules/ui/libs/hierarchy.js]: Please extract this block to function - `moveRootToFront` / `ensureRootIsFirst` or something

File: lib/ui/src/modules/ui/libs/hierarchy.js
Comment (igor-dv) [reply lib/ui/src/modules/ui/libs/hierarchy.js]: 👌

**Review by igor-dv — APPROVED** [review]
Very nice tests! + tested locally.

File: lib/ui/src/modules/ui/libs/hierarchy.js
Comment (Havelaer) [reply lib/ui/src/modules/ui/libs/hierarchy.js]: Extracted to `ensureMainRootIsFirst` and used `Object.values` instead of `Object.keys(...).map(...)`

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
Good to go if you ask me!

**Comment by danielduan** [comment]
I think we're aiming to add this for `3.4`?

I added the `do not merge` label because of the upcoming changes to our workflow. Feel free to merge to master after we get `3.3` out

**Comment by Hypnosphi** [comment]
We have moved some stories to `examples/storybook-official`. Would you mind resolving the conflicts?

**Review by ndelangen — APPROVED** [review]

**Comment by Hypnosphi** [comment]
@shilman after we merge this one, we won't be able to release `3.3` from `master` anymore

**Comment by Havelaer** [comment]
I've rebased the branch on master, keeping the history nice & clean. Fixed tests, added separator config to examples/official-storybook and updated core and cli snapshots.

**Comment by ndelangen** [comment]
@Havelaer That's fantastic work!

**Review by ndelangen — APPROVED** [review]

**Comment by igor-dv** [comment]
@Havelaer maybe you can resolve the conflicts and merge?

**Comment by ndelangen** [comment]
I resolved merge conflicts

**Comment by igor-dv** [comment]
👏

File: addons/options/README.md
Comment (viddo) [reply addons/options/README.md]: Could the PR description be updated to have it in the example there? Didn't work on the first try and had to search around to discover this.

Found my way here through the [3.4.0-alpha0 changelog entry](https://github.com/storybooks/storybook/blob/next/CHANGELOG.md#340-alpha0)

**Pull request merged**
